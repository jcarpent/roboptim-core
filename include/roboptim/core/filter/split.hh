// Copyright (C) 2010 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_FILTER_SPLIT_HH
# define ROBOPTIM_CORE_FILTER_SPLIT_HH
# include <stdexcept>
# include <boost/shared_ptr.hpp>

# include <roboptim/core/n-times-derivable-function.hh>

namespace roboptim
{
  /// \addtogroup roboptim_filter
  /// @{

  /// \brief Select an element of a function's output.
  /// \tparam T input function type.
  template <typename T>
  class Split : public T
  {
  public:
    /// \brief Import value type.
    typedef typename DifferentiableFunction::value_type value_type;
    /// \brief Import size type.
    typedef typename DifferentiableFunction::size_type size_type;
    /// \brief Import vector type.
    typedef typename DifferentiableFunction::vector_t vector_t;
    /// \brief Import result type.
    typedef typename DifferentiableFunction::result_t result_t;
    /// \brief Import argument type.
    typedef typename DifferentiableFunction::argument_t argument_t;
    /// \brief Import gradient type.
    typedef typename DifferentiableFunction::gradient_t gradient_t;
    /// \brief Import hessian type.
    typedef typename TwiceDifferentiableFunction::hessian_t hessian_t;
    /// \brief Import jacobian type.
    typedef typename DifferentiableFunction::jacobian_t jacobian_t;
    /// \brief Import interval type.
    typedef typename DifferentiableFunction::interval_t interval_t;

    /// \brief Split filter constructor.
    /// \param fct input function.
    /// \param functionId index of the output to select.
    explicit Split (boost::shared_ptr<const T> fct,
		    size_type functionId);
    ~Split ();

  protected:
    virtual void impl_compute (result_t& result, const argument_t& argument)
      const;


    virtual void impl_gradient (gradient_t& gradient,
				const argument_t& argument,
				size_type functionId = 0)
      const;

    virtual void impl_hessian (hessian_t& hessian,
    			       const argument_t& argument,
    			       size_type functionId = 0) const;

    virtual void impl_derivative (gradient_t& derivative,
    				  value_type argument,
    				  size_type order = 1) const;

  private:
    boost::shared_ptr<const T> function_;
    size_type functionId_;
    mutable result_t res_;
  };

  template <typename P, typename C>
  void addNonScalarConstraint
  (P& problem,
   boost::shared_ptr<C> constraint,
   std::vector<Function::interval_t> interval,
   std::vector<Function::value_type> scale
   = std::vector<Function::value_type> ());

  /// @}

} // end of namespace roboptim

# include <roboptim/core/filter/split.hxx>
#endif //! ROBOPTIM_CORE_FILTER_SPLIT_HH
