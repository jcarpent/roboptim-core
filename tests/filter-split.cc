// Copyright (C) 2010 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#undef NDEBUG

#include "shared-tests/fixture.hh"

#include <iostream>

#include <roboptim/core/io.hh>
#include <roboptim/core/differentiable-function.hh>
#include <roboptim/core/util.hh>
#include <roboptim/core/filter/split.hh>

using namespace roboptim;

struct F : public DifferentiableFunction
{
  F () : DifferentiableFunction (1, 10, "f_n (x) = n * x")
  {}

  void impl_compute (result_t& res, const argument_t& argument) const
  {
    res.setZero ();
    for (size_type i = 0; i < outputSize (); ++i)
      res[i] = (value_type)i * argument[0];
  }

  void impl_gradient (gradient_t& grad, const argument_t&,
		      size_type functionId) const
  {
    grad.setZero ();
    grad[0] = (value_type)functionId;
  }
};

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE (split)
{
  boost::shared_ptr<boost::test_tools::output_test_stream>
    output = retrievePattern ("filter-split");

  boost::shared_ptr<F> f (new F ());

  for (DifferentiableFunction::vector_t::Index id = 0; id < 10; ++id)
    {
      Split<DifferentiableFunction> splitF (f, id);

      (*output) << splitF << ":" << std::endl
		<< std::endl;

      Function::vector_t x (1);
      for (double i = 0.; i < 10.; i += 0.5)
	{
	  x[0] = i;
	  (*output) << "x = " << x[0] << std::endl;
	  (*output) << splitF (x) << std::endl;
	  (*output) << splitF.gradient (x) << std::endl;
	  (*output) << std::endl;
	}
    }

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()
