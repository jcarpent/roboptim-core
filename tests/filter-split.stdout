f_n (x) = n * x (split, function Id = 0) (differentiable function):

x = 0
[1](0)
[1](0)

x = 0.5
[1](0)
[1](0)

x = 1
[1](0)
[1](0)

x = 1.5
[1](0)
[1](0)

x = 2
[1](0)
[1](0)

x = 2.5
[1](0)
[1](0)

x = 3
[1](0)
[1](0)

x = 3.5
[1](0)
[1](0)

x = 4
[1](0)
[1](0)

x = 4.5
[1](0)
[1](0)

x = 5
[1](0)
[1](0)

x = 5.5
[1](0)
[1](0)

x = 6
[1](0)
[1](0)

x = 6.5
[1](0)
[1](0)

x = 7
[1](0)
[1](0)

x = 7.5
[1](0)
[1](0)

x = 8
[1](0)
[1](0)

x = 8.5
[1](0)
[1](0)

x = 9
[1](0)
[1](0)

x = 9.5
[1](0)
[1](0)

f_n (x) = n * x (split, function Id = 1) (differentiable function):

x = 0
[1](0)
[1](1)

x = 0.5
[1](0.5)
[1](1)

x = 1
[1](1)
[1](1)

x = 1.5
[1](1.5)
[1](1)

x = 2
[1](2)
[1](1)

x = 2.5
[1](2.5)
[1](1)

x = 3
[1](3)
[1](1)

x = 3.5
[1](3.5)
[1](1)

x = 4
[1](4)
[1](1)

x = 4.5
[1](4.5)
[1](1)

x = 5
[1](5)
[1](1)

x = 5.5
[1](5.5)
[1](1)

x = 6
[1](6)
[1](1)

x = 6.5
[1](6.5)
[1](1)

x = 7
[1](7)
[1](1)

x = 7.5
[1](7.5)
[1](1)

x = 8
[1](8)
[1](1)

x = 8.5
[1](8.5)
[1](1)

x = 9
[1](9)
[1](1)

x = 9.5
[1](9.5)
[1](1)

f_n (x) = n * x (split, function Id = 2) (differentiable function):

x = 0
[1](0)
[1](2)

x = 0.5
[1](1)
[1](2)

x = 1
[1](2)
[1](2)

x = 1.5
[1](3)
[1](2)

x = 2
[1](4)
[1](2)

x = 2.5
[1](5)
[1](2)

x = 3
[1](6)
[1](2)

x = 3.5
[1](7)
[1](2)

x = 4
[1](8)
[1](2)

x = 4.5
[1](9)
[1](2)

x = 5
[1](10)
[1](2)

x = 5.5
[1](11)
[1](2)

x = 6
[1](12)
[1](2)

x = 6.5
[1](13)
[1](2)

x = 7
[1](14)
[1](2)

x = 7.5
[1](15)
[1](2)

x = 8
[1](16)
[1](2)

x = 8.5
[1](17)
[1](2)

x = 9
[1](18)
[1](2)

x = 9.5
[1](19)
[1](2)

f_n (x) = n * x (split, function Id = 3) (differentiable function):

x = 0
[1](0)
[1](3)

x = 0.5
[1](1.5)
[1](3)

x = 1
[1](3)
[1](3)

x = 1.5
[1](4.5)
[1](3)

x = 2
[1](6)
[1](3)

x = 2.5
[1](7.5)
[1](3)

x = 3
[1](9)
[1](3)

x = 3.5
[1](10.5)
[1](3)

x = 4
[1](12)
[1](3)

x = 4.5
[1](13.5)
[1](3)

x = 5
[1](15)
[1](3)

x = 5.5
[1](16.5)
[1](3)

x = 6
[1](18)
[1](3)

x = 6.5
[1](19.5)
[1](3)

x = 7
[1](21)
[1](3)

x = 7.5
[1](22.5)
[1](3)

x = 8
[1](24)
[1](3)

x = 8.5
[1](25.5)
[1](3)

x = 9
[1](27)
[1](3)

x = 9.5
[1](28.5)
[1](3)

f_n (x) = n * x (split, function Id = 4) (differentiable function):

x = 0
[1](0)
[1](4)

x = 0.5
[1](2)
[1](4)

x = 1
[1](4)
[1](4)

x = 1.5
[1](6)
[1](4)

x = 2
[1](8)
[1](4)

x = 2.5
[1](10)
[1](4)

x = 3
[1](12)
[1](4)

x = 3.5
[1](14)
[1](4)

x = 4
[1](16)
[1](4)

x = 4.5
[1](18)
[1](4)

x = 5
[1](20)
[1](4)

x = 5.5
[1](22)
[1](4)

x = 6
[1](24)
[1](4)

x = 6.5
[1](26)
[1](4)

x = 7
[1](28)
[1](4)

x = 7.5
[1](30)
[1](4)

x = 8
[1](32)
[1](4)

x = 8.5
[1](34)
[1](4)

x = 9
[1](36)
[1](4)

x = 9.5
[1](38)
[1](4)

f_n (x) = n * x (split, function Id = 5) (differentiable function):

x = 0
[1](0)
[1](5)

x = 0.5
[1](2.5)
[1](5)

x = 1
[1](5)
[1](5)

x = 1.5
[1](7.5)
[1](5)

x = 2
[1](10)
[1](5)

x = 2.5
[1](12.5)
[1](5)

x = 3
[1](15)
[1](5)

x = 3.5
[1](17.5)
[1](5)

x = 4
[1](20)
[1](5)

x = 4.5
[1](22.5)
[1](5)

x = 5
[1](25)
[1](5)

x = 5.5
[1](27.5)
[1](5)

x = 6
[1](30)
[1](5)

x = 6.5
[1](32.5)
[1](5)

x = 7
[1](35)
[1](5)

x = 7.5
[1](37.5)
[1](5)

x = 8
[1](40)
[1](5)

x = 8.5
[1](42.5)
[1](5)

x = 9
[1](45)
[1](5)

x = 9.5
[1](47.5)
[1](5)

f_n (x) = n * x (split, function Id = 6) (differentiable function):

x = 0
[1](0)
[1](6)

x = 0.5
[1](3)
[1](6)

x = 1
[1](6)
[1](6)

x = 1.5
[1](9)
[1](6)

x = 2
[1](12)
[1](6)

x = 2.5
[1](15)
[1](6)

x = 3
[1](18)
[1](6)

x = 3.5
[1](21)
[1](6)

x = 4
[1](24)
[1](6)

x = 4.5
[1](27)
[1](6)

x = 5
[1](30)
[1](6)

x = 5.5
[1](33)
[1](6)

x = 6
[1](36)
[1](6)

x = 6.5
[1](39)
[1](6)

x = 7
[1](42)
[1](6)

x = 7.5
[1](45)
[1](6)

x = 8
[1](48)
[1](6)

x = 8.5
[1](51)
[1](6)

x = 9
[1](54)
[1](6)

x = 9.5
[1](57)
[1](6)

f_n (x) = n * x (split, function Id = 7) (differentiable function):

x = 0
[1](0)
[1](7)

x = 0.5
[1](3.5)
[1](7)

x = 1
[1](7)
[1](7)

x = 1.5
[1](10.5)
[1](7)

x = 2
[1](14)
[1](7)

x = 2.5
[1](17.5)
[1](7)

x = 3
[1](21)
[1](7)

x = 3.5
[1](24.5)
[1](7)

x = 4
[1](28)
[1](7)

x = 4.5
[1](31.5)
[1](7)

x = 5
[1](35)
[1](7)

x = 5.5
[1](38.5)
[1](7)

x = 6
[1](42)
[1](7)

x = 6.5
[1](45.5)
[1](7)

x = 7
[1](49)
[1](7)

x = 7.5
[1](52.5)
[1](7)

x = 8
[1](56)
[1](7)

x = 8.5
[1](59.5)
[1](7)

x = 9
[1](63)
[1](7)

x = 9.5
[1](66.5)
[1](7)

f_n (x) = n * x (split, function Id = 8) (differentiable function):

x = 0
[1](0)
[1](8)

x = 0.5
[1](4)
[1](8)

x = 1
[1](8)
[1](8)

x = 1.5
[1](12)
[1](8)

x = 2
[1](16)
[1](8)

x = 2.5
[1](20)
[1](8)

x = 3
[1](24)
[1](8)

x = 3.5
[1](28)
[1](8)

x = 4
[1](32)
[1](8)

x = 4.5
[1](36)
[1](8)

x = 5
[1](40)
[1](8)

x = 5.5
[1](44)
[1](8)

x = 6
[1](48)
[1](8)

x = 6.5
[1](52)
[1](8)

x = 7
[1](56)
[1](8)

x = 7.5
[1](60)
[1](8)

x = 8
[1](64)
[1](8)

x = 8.5
[1](68)
[1](8)

x = 9
[1](72)
[1](8)

x = 9.5
[1](76)
[1](8)

f_n (x) = n * x (split, function Id = 9) (differentiable function):

x = 0
[1](0)
[1](9)

x = 0.5
[1](4.5)
[1](9)

x = 1
[1](9)
[1](9)

x = 1.5
[1](13.5)
[1](9)

x = 2
[1](18)
[1](9)

x = 2.5
[1](22.5)
[1](9)

x = 3
[1](27)
[1](9)

x = 3.5
[1](31.5)
[1](9)

x = 4
[1](36)
[1](9)

x = 4.5
[1](40.5)
[1](9)

x = 5
[1](45)
[1](9)

x = 5.5
[1](49.5)
[1](9)

x = 6
[1](54)
[1](9)

x = 6.5
[1](58.5)
[1](9)

x = 7
[1](63)
[1](9)

x = 7.5
[1](67.5)
[1](9)

x = 8
[1](72)
[1](9)

x = 8.5
[1](76.5)
[1](9)

x = 9
[1](81)
[1](9)

x = 9.5
[1](85.5)
[1](9)

